### PR TITLE
[Core] Add HTML reprs for `ClientContext` and `WorkerContext`

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

These changes are the first in a series intended to improve integration with notebooks. They establish the directory structure and dependencies required to render the richer output that notebooks support.

## Summary

* This PR adds a simple `_html_repr_` for `ClientContext` and `WorkerContext`, which are returned from `ray.init()`. Now when a user initializes Ray in a notebook, they're greeted with a neat little table of information about the context, and a clickable link to the dashboard:

image

* Added Jinja2 >= 3.0.3 as a dependency for rendering HTML templates in the notebook. 3.0.3 is the last Jinja2 version to support Python 3.6, so whenever Ray drops support for 3.6 we can bump this.
* Templates are stored in `ray/python/ray/widgets/templates/`
* I haven't made any modifications to how logging is handled in the notebook. Maybe it would be good to redirect logging in the notebook into a separate pane so that the user can view/search it easily? We might need a discussion about how to do this so that duplicate information isn't displayed here in the future. For now, I've left it as is.

## Related issue number

See #25591 for more information.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

**Note**: Looks like there's an issue installing the `Jinja2 >= 3.0.3` dependency on CI, though I have checked that it is available for Python 3.6 from PyPI. Is there some other change that needs to be made to get CI to pass?

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -290,6 +290,7 @@ def get_packages(self):
         "frozenlist",
         "requests",
         "virtualenv",  # For pip runtime env.
+        "Jinja2 >= 3.0.3",
```
Comment (richardliaw) [inline python/setup.py:293]: could we only take this dep in ray[default]?

**Review by richardliaw — CHANGES_REQUESTED** [review]
this is great! just worried about dependency growth here.

File: python/setup.py (hunk 1 above)
Comment (peytondmurray) [reply python/setup.py]: For sure!

**Comment by pcmoritz** [comment]
@peytondmurray @richardliaw 

I have a different suggestion to avoid the dependency on `Jinja2`: Why don't we just fill in the values by doing string substitutions of the values? We don't really need the advanced jinja2 capabilities here and if we just do it with strings, we can still have consistent behavior for `ray` and `ray[default]` :)

We can still keep the same syntax, so if we need to switch to jinja2 in the future, that would be easy.

**Comment by peytondmurray** [comment]
@pcmoritz This sounds perfectly fine to me, I'll remove the `Jinja2` dependency now.

**Comment by peytondmurray** [comment]
@pcmoritz Hmm, I spoke too soon in the previous comment. In this PR we are already using more than just Jinja2's variable substitution. We also make use of `{% if <variable> %} ... {% endif %}`, which might be more cumbersome to start implementing. We also already have `Jinja2` as a dependency for some release tools:

* https://github.com/ray-project/ray/blob/97582a802d4519ae72120464bf9c1c6333eec4bc/release/requirements_buildkite.txt#L4
* https://github.com/ray-project/ray/blob/2cf20e5406d77801925c3210fff23dbc6f8ba02f/release/ray_release/template.py#L7

If you still think it's better to avoid adding `Jinja2` as dependency, we can probably find a way to get by without it, though. What do you think?

**Comment by richardliaw** [comment]
Yeah, I think we should try really hard to avoid the dependency. The release infra takes on the dependency because it's not user facing; more worried about introducing dependencies at the core end-user level

**Comment by peytondmurray** [comment]
Looks like the CI is stuck. For now, I'll re-request review and if you'd like me to push to rerun the stuck tests, let me know.

**Comment by peytondmurray** [comment]
Now that the tests have finished:

* `brain-rllib-learning-cont-actions-pytorch`: `//rllib:learning_tests_pendulum_cql` [is flaky](https://flakey-tests.ray.io/) and has timed out
* `kubernetes-operator`: Failed to set up a test environment
* `mac wheels and jars`: Very flaky test recently

so I think these are unrelated. @richardliaw If this all looks okay to you, I think this is ready to go!

**Comment by pcmoritz** [comment]
@richardliaw Can you review this PR? Peyton is going on vacation next week and we should try to merge it before that :)

File: python/ray/widgets/render.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    def __init__(self, file: str):
+        with open(pathlib.Path(__file__).parent / "templates" / file, "r") as f:
+            self.template = f.read()
+
+    def render(self, **kwargs) -> str:
+        """Render a template by replacing instances of `{{ key }}` with `value`
+        from the keyword arguments.
+
+        Returns
```
Comment (richardliaw) [inline python/ray/widgets/render.py:19]: We use google style python docstrings btw, could we edit the style of the docstrings to match?

**Review by richardliaw — CHANGES_REQUESTED** [review]
just a couple nits!

File: python/ray/widgets/__init__.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+from .render import Template
```
[the remark this answers is not in the document]
Comment (peytondmurray) [reply python/ray/widgets/__init__.py:1]: :100:

File: python/ray/widgets/render.py (hunk 2 above)
Comment (peytondmurray) [reply python/ray/widgets/render.py:19]: :facepalm: sorry about this! I've reformatted the docstrings accordingly!

**Review by richardliaw — APPROVED** [review]

**Pull request merged**